fawltydeps/gitignore_parser.py

```diff
@@ -0,0 +1,280 @@
+import os
+import re
+import sys
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    NamedTuple,
+    Optional,
+    Tuple,
+)
+
+from fawltydeps.types import Location
+
+if TYPE_CHECKING or sys.version_info >= (3, 9):
+    CompiledRegex = re.Pattern[str]  # pylint: disable=unsubscriptable-object
+else:  # re.Pattern not subscriptable before Python 3.9
+    CompiledRegex = re.Pattern
+
+logger = logging.getLogger(__name__)
+
+
+class IgnoreRuleError(ValueError):
+    """Error while parsing an ignore pattern in DirectoryTraversal.ignore()."""
+
+    def __init__(self, msg: str, pattern: str, source: Optional[Location] = None):
+        self.msg = msg
+        self.pattern = pattern
+        self.source = source
+
+    def __str__(self) -> str:
+        if self.source is not None:
+            return f"{self.msg}: {self.pattern!r} ({self.source})"
+        return f"{self.msg}: {self.pattern!r}"
+
+
+class IgnoreRuleMissing(IgnoreRuleError):
+    """A blank line or comment passed to DirectoryTraversal.ignore()."""
+
+
+def parse_gitignore(
+    full_path: Path, base_dir: Optional[Path] = None
+) -> Callable[[Path, bool], bool]:
+    """Parse the given .gitignore file and return the resulting rules checker.
+
+    The 'base_dir', if given, specifies the directory relative to which the
+    parsed ignore rules will be interpreted. If not given, the parent directory
+    of the 'full_path' is used instead.
+
+    See parse_gitignore_lines() for more details.
+    """
+    if base_dir is None:
+        base_dir = full_path.parent
+    with open(full_path) as ignore_file:
+        return parse_gitignore_lines(ignore_file, base_dir, full_path)
+
+
+def parse_gitignore_lines(
+    lines: Iterable[str],
+    base_dir: Optional[Path] = None,
+    file_hint: Optional[Path] = None,
+) -> Callable[[Path, bool], bool]:
+    """Parse gitignore lines and return the resulting rules checker.
+
+    The return value is a predicate for paths to ignore, i.e. a callable that
+    returns True/False based on whether a given path should be ignored or not.
+
+    The 'base_dir', if given, specifies the directory relative to which the
+    parsed ignore rules will be interpreted. If not given, the given rules
+    cannot be anchored.
+    """
+    rules = []
+    for lineno, line in enumerate(lines, start=1):
+        source = None if file_hint is None else Location(file_hint, lineno=lineno)
+        line = line.rstrip("\n")
+        try:
+            rules.append(IgnoreRule.from_pattern(line, base_dir, source))
+        except IgnoreRuleMissing as exc:
+            # Blank lines and comments are ok when parsing multiple lines
+            logger.debug(str(exc))
+
+    return functools.partial(match_rules, rules)
+
+
+def match_rules(rules: List["IgnoreRule"], path: Path, is_dir: bool) -> bool:
+    """Match the given path against the given list of rules."""
+    for rule in reversed(rules):
```

Comment:
Consider a negated rule like this:

```
foo.*
!foo.py
```

that is, we want to ignore `foo.*`, except for `foo.py`.

Here is the [original code in gitignore-parser](https://github.com/mherrmann/gitignore_parser/blob/6c8bd44695546269b0d1f8f41819fe480baea954/gitignore_parser.py#L9) where this came from:

```python
def handle_negation(file_path, rules: Reversible["IgnoreRule"]):
    for rule in reversed(rules):
        if rule.match(file_path):
            return not rule.negation
    return False

def parse_gitignore(full_path, base_dir=None):
    ...
    if not any(r.negation for r in rules):
        return lambda file_path: any(r.match(file_path) for r in rules)
    else:
        # We have negation rules. We can't use a simple "any" to evaluate them.
        # Later rules override earlier rules.
        return lambda file_path: handle_negation(file_path, rules)
```

So in the case where there are no negated rules, there is no difference: For a given path, we can process the ignore rules in any order: as long as any of them match, we will ignore the path.

But for negated rules, we want later rules to override earlier rules, hence we want to process them in the reverse order, so that we can stop processing as soon as the first match is found (i.e. we want to `foo.py` to match `!foo.py`, _before_ it matches `foo.*`).

When refactoring this, I decided to simplify the code to always do this reverse processing. (The way lists are implemented in Python, I don't expect a reverse traversal to take much more time in the common case where there are no negated patterns)

Reply:
> But for negated rules, we want later rules to override earlier rules, hence we want to process them in the reverse order, so that we can stop processing as soon as the first match is found (i.e. we want to foo.py to match !foo.py, before it matches foo.*).

Thank you for the explanation. Worth to remember as well that an order in which you put entries in `.gitignore` file is important. (TILA - today I learned again :laughing: )